exercise-book/src/calculator.md

```diff
@@ -0,0 +1,700 @@
+
+* Let's make `match` produce an expression: `let expr = match word { ... };`
+* Let's change our `stack` variable to store `Box`ed expressions
+
+```rust ignore
+pub fn parse(input: &str) -> Result<Expr, ParseError> {
+    let mut stack: Vec<Box<Expr>> = Vec::new();
+    for word in input.split_ascii_whitespace() {
+        let expr = match word {
+            "+" => {
+                let a = stack.pop().ok_or(ParseError::MissingOperand)?;
+                let b = stack.pop().ok_or(ParseError::MissingOperand)?;
+                Expr::Add(a, b)
+            }
+            "sqr" => {
+                let a = stack.pop().ok_or(ParseError::MissingOperand)?;
+                Expr::Square(a)
+            }
+            _ => {
+                let number = word
+                    .parse::<i64>()
+                    .map_err(|_| ParseError::NotANumber(word.to_string()))?;
+                Expr::Number(number)
+            }
+        };
+        stack.push(Box::new(expr));
+    }
+    assert_eq!(stack.len(), 1);
+    let res = stack.pop().unwrap();
+    Ok(*res)
+}
+```
+
+Note that we have to adjust the type from `Box<Expr>` to `Expr` at the end.
+While we are at it let's rewrite the end portion of the function to get rid of an assertion and an `unwrap`:
+
+```rust ignore
+for word in input.split_ascii_whitespace() {
+    // ...
+}
+
+match stack.pop() {
+    Some(expr) if stack.is_empty() => Ok(*expr),
+    Some(_) => Err(ParseError::TooManyOperands),
+    None => Err(ParseError::EmptyInput),
+}
+```
+
+Let's test our new error conditions:
+
+```rust ignore
+#[test]
+fn too_many_operands() {
+    let input = "42 42 42 +";
+    let error = parse(input).unwrap_err();
+    assert_eq!(error, ParseError::TooManyOperands);
+}
+
+#[test]
+fn empty_input() {
+    let input = "      ";
+    let error = parse(input).unwrap_err();
+    assert_eq!(error, ParseError::EmptyInput);
+}
+```
+
+#### `eval` function
+
+So far every branch of the `match` produces an number that we later wrap into `Ok`.
+
+```rust ignore
+pub fn eval(expr: &Expr) -> Result<i64, EvalError> {
+    let value = match expr {
+        Expr::Number(n) => *n,
+        Expr::Add(a, b) => eval(a)? + eval(b)?,
+        Expr::Square(a) => eval(a)?.pow(2),
+    };
+    Ok(value)
+}
+```
+
+### Step 6: Subtraction
+
+In `parse` function all subsequent operators will require two `stack.pop()` calls.
+We may as well combine different operators together and use `unreachable!` macro for the second match:
+
+```rust ignore
+match word {
+    "+" | "-" => {
+        let a = stack.pop().ok_or(ParseError::MissingOperand)?;
+        let b = stack.pop().ok_or(ParseError::MissingOperand)?;
+        match word {
+            "+" => Expr::Add(a, b),
+            "-" => Expr::Sub(a, b),
+            _ => unreachable!(),
+        }
+    }
+    // ...
+}
+```
+
+```rust ignore
+// in `eval`
+match expr {
+    // ...
+    Expr::Add(a, b) => eval(a)? + eval(b)?,
+    Expr::Sub(a, b) => eval(a)? - eval(b)?,
```

Comment:
Should we talk about `checked_add` and `checked_sub` and how arithmetic can panic and so shouldn't be performed on untrusted inputs?

Reply:
Probably? When I'm running this exercise with groups of different backgrounds everyone remembers the division by zero, but only people with some C or embedded background mention overflows (in three years it only happened once!). So, I keep it light and simple with operators. 

Handling overflows can be a stretch goal, and I can mention it in the exercise. Good call!